docs/readmes/proposals/p017_mme_migrate_to_c++.md

```diff
@@ -0,0 +1,133 @@
+- S1AP, S1AP_IMSI_MAP, S1AP_UE
+- MME_UE, MME_NAS
+- SGW, SGW_UE, SPGW, SPGW_UE
+- S11/GTPV2-C
+- NGAP, NGAP_IMSI_MAP, NGAP_UE
+
+#### S1AP, MME_APP, NAS migration
+
+ S1AP seems to have no state dependency with other tasks.
+ 
+ Should be an opportunity to start doing a first migration with S1AP.
+
+ MME_APP and NAS tasks have contexts in common.
+ 
+
+////### Migrate some MM and NAS IEs
+////For the ease of manipulating IEs, migrate a selected list of IEs (operator ==, copy, conversion, etc)
+
+
+### 3GPP protocols (OAI)
+This work is necessary only for the deployment of Magma MME with CUPS architecture.
+
+S6a Interface implemented in C should be migrated to C++.
+A C++ port of S6a interface is already available in OAI HSS repo (https://github.com/OPENAIRINTERFACE/openair-hss) could serve as a basis.
+
+S11 Interface and its GTPV2-C protocol implemented in C should be migrated to C++.
+A C++ port of S11 and GTPV2-C interface done in OAI SPGW-C repo (https://github.com/OPENAIRINTERFACE/openair-spgwc) could serve as a basis.
+
+
+## Rationale
+
+### Get/Set state attributes
+
+For a progressive migration, and a decoupling of MME core code from the serialization library, macros, inline, helpers functions/C++ patterns may be used instead of directly using serialization library API in MME core code.
+
+Some MM and NAS IEs, for the ease of their manipulation (comparison operators, etc) could be migrated with wrapper classes.
+
+[A discussion of alternate approaches and the trade offs, advantages, and
+disadvantages of the specified approach.]
+
+## Compatibility
+
+Endianness: Whatever is the HW platform, endianness is fixed.
+
+Backward compatibility of the new serialization library with protobuf is not envisioned.
+
+## Observability and Debug
+
+TODO
+[A description, of how issues with this design would be observed and debugged
+in various stages from development through production.]
+
+## Implementation
+
+  Description of steps in the C++ migration:
+ 
+* Serialization library selection
+    * Build 2 realistics Prototypes with MME_APP, NAS protos (cap’n proto, Flat buffers)
+        * Migrate hashtables (find functions)
+    * Get measurements and evaluate performances, maintainability.
+    * Select library
+* MME_APP/NAS Serialization migration
+    * Write proto
+    * Write redis client libraries
+    * Design classes for wrapping generated POD classes (?)
+    	Not all context attribute need to be serialized
+    * Migrate states (update core code with C++ patterns)
+    * Write Unit tests
+* Migrate MME_APP/NAS hashtables (find functions, evaluate performance)
+* S1AP Serialization migration
+    * Write proto
+    * Write redis client libraries
+    * Design classes for wrapping generated POD classes 
+    	Not all context attribute need to be serialized
+    * Migrate states (update core code)
+    * Write Unit tests
+* Migrate S1AP hashtables (find functions, evaluate performance)
```

Comment:
nit: Can we specify all MME task threads (S1AP, MME and SPGW) state hashtables?

Reply:
working on it...

Reply:
I made an update, let me know.

Reply:
This looks good to me, thanks!